contracts/cw3-multiple-choice/src/msg.rs

```diff
@@ -0,0 +1,165 @@
+    ReverseProposals {
+        start_before: Option<u64>,
+        limit: Option<u32>,
+    },
+    /// Returns the number of proposals in the DAO (u64)
+    ProposalCount {},
+    // Returns config
+    GetConfig {},
+    /// Returns VoteResponse
+    Vote {
+        proposal_id: u64,
+        voter: String,
+    },
+    /// Returns VoteListResponse
+    ListVotes {
+        proposal_id: u64,
+        start_after: Option<String>,
+        limit: Option<u32>,
+    },
+    /// Returns information about current tallys for a
+    /// proposal. Returns type `VoteTallyResponse`.
+    Tally {
+        proposal_id: u64,
+    },
+    /// Returns VoterInfo
+    Voter {
+        address: String,
+    },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct VoteMsg {
+    pub proposal_id: u64,
+    pub vote: Vote,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct ProposeMsg {
+    pub title: String,
+    pub description: String,
+    pub choices: Vec<String>,
+    pub msgs: Vec<Vec<CosmosMsg<Empty>>>,
+    // note: we ignore API-spec'd earliest if passed, always opens immediately
+    pub latest: Option<Expiration>,
+}
+
+/// This defines the different ways tallies can happen.
+///
+/// The total_weight used for calculating success as well as the weights of each
+/// individual voter used in tallying should be snapshotted at the beginning of
+/// the block at which the proposal starts (this is likely the responsibility of a
+/// correct cw4 implementation).
+/// See also `ThresholdResponse` in the cw3 spec.
+/// This is similar to thresholds for 'yes' votes in single choice voting, but in this case in order for any
+/// of the multiple choice options to pass, it must have threshold votes in its favor.
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+#[serde(rename_all = "snake_case")]
+pub enum Threshold {
+    /// Declares a percentage of the total weight that must cast a vote for any given option in order for
+    /// a proposal to be executed.
+    AbsolutePercentage { percentage: Decimal },
+
+    /// Declares a `quorum` of the total votes that must participate in the election in order
+    /// for the vote to be considered at all. Within that quorum, threshold must vote in favor of
+    /// any given option.
+    ThresholdQuorum { threshold: Decimal, quorum: Decimal },
+}
+
+impl Threshold {
+    /// returns error if this is an unreachable value,
+    /// given a total weight of all members in the group
+    pub fn validate(&self) -> Result<(), ContractError> {
+        match self {
+            Threshold::AbsolutePercentage {
+                percentage: percentage_needed,
+            } => valid_percentage(percentage_needed),
+            Threshold::ThresholdQuorum {
+                threshold,
+                quorum: quroum,
+            } => {
+                valid_percentage(threshold)?;
+                valid_percentage(quroum)
```

Comment:
We likely want to support zero percent quorums (allows purely plurality voting like US elections). For `cw-govmod-single` I added a `validate_quorum` function to handle this.